### PR TITLE
make figure pip installable

### PR DESCRIPTION
**Pull request by aleksandra-tarkowska**
# What this PR does

pip install figure

# Testing this PR

```
$ mkdir test
$ cd test
$ virtualenv venvweb
$ source venvweb/bin/activate
(venvweb) $ pip install omego
(venvweb) $ omego download --ice 3.6 --branch OMERO-DEV-merge-build py
(venvweb) $ mv OMERO.py[currentversion] OMERO.py
(venvweb) $ pip install -r OMERO.py/share/web/requirements-py27-trial.txt

(venvweb) $ pip install git+git://github.com/aleksandra-tarkowska/figure.git@pip_installer
(venvweb) $ OMERO.py/bin/omero config append omero.web.apps '"figure"'
(venvweb) $ OMERO.py/bin/omero config append omero.web.ui.top_links '["Figure", "figure_index", {"title": "Open Figure in new tab", "target": "figure"}]' 
(venvweb) $ OMERO.py/bin/omero web start
(venvweb) $ cd venvweb/lib/python2.7/site-packages/figure/scripts
(venvweb) $ ~/test/OMERO.py/bin/omero script upload omero/figure_scripts/Figure_To_Pdf.py --official 
```

then generate nginx config and reload. Web should operate as normal

Full doc in http://www.openmicroscopy.org/site/support/omero5.3-staging/sysadmins/unix/install-web/install-web-trial.html

see https://github.com/openmicroscopy/openmicroscopy/pull/4808 for reference

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -75,3 +82,14 @@ If you are using docker-machine (e.g. on Mac OS X or Windows), you can find the
     192.168.99.100
     # Now check the result in your browser at:
     http://192.168.99.100:8000/
+
+
+License
+-------
+
+MAPR is released under the AGPL.
```
Comment (sbesson) [inline README.rst]: OMERO.figure

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+setup(name="omero-figure",
+      packages=find_packages(exclude=['ez_setup']),
+      version=VERSION,
+      description="A Python plugin for OMERO.web",
+      long_description=read('README.rst'),
+      author='The Open Microscopy Team',
+      author_email='ome-devel@lists.openmicroscopy.org.uk',
+      license='AGPLv3',
+      url="https://github.com/ome/figure",
+      download_url='https://github.com/ome/figure/tarball/1.2.2',  # NOQA
```
Comment (sbesson) [inline setup.py]: This could use the substitution of `VERSION`

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      keywords=['OMERO.web', 'figure'],
+      install_requires=['omego', 'reportlab', 'markdown'],
```
Comment (sbesson) [inline setup.py]: Worth discussing. I understand the target is to make wider usage of `omego` for the deployment but I am not certain it qualifies as a *minimal* requirement of OMERO.figure - see https://packaging.python.org/requirements/

**Comment by sbesson** [comment]
On the licensing front, I understand our preference should be to AGPL licensing for all OME-maintained Web apps. Is this generally a policy we should apply to all our repos? /cc @joshmoore Note there are still a few places in the code mentioning the traditional GPL (e.g. package.json) which need to be relicensed.

**Comment by aleksandra-tarkowska** [comment]
Done

**Comment by aleksandra-tarkowska** [comment]
I added requirements.txt. Does it need a version caps, it wasn't clear from the figure doc exactly?

**Comment by joshmoore** [comment]
@will-moore : what's the potential for moving to this layout in a point release? Will anything break?

**Comment by will-moore** [comment]
@joshmoore It might seem confusing to our users to release a new version with (almost) the only new feature being that it's installed in a different way. Why would you install it if it has no other new features?
It will cost ~a day of release prep etc and break all currently open PRs. Is it worth it?

**Comment by joshmoore** [comment]
Sorry, I wasn't clear. I meant, can this be a part of the next patch release? Or does it need something bigger/breaking? Once this is merged, we can point to github for merge pip installs. The question of the open PRs is a good one, and perhaps a better way of phrasing my question is: when can _this_ be merged?

**Comment by will-moore** [comment]
@aleksandra-tarkowska You need to add "figure/" to all the template paths. Here's a commit for that:
https://github.com/will-moore/figure/commit/9b930ea0ebd5a1196a801ec0bd715d2bff5a24e3

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import os
+from setuptools import setup, find_packages
+
+
+# Utility function to read the README file.
+# Used for the long_description.  It's nice, because now 1) we have a top level
+# README file and 2) it's easier to type in the README file than to put a raw
+# string in below ...
+def read(fname):
+    return open(os.path.join(os.path.dirname(__file__), fname)).read()
```
Comment (will-moore) [inline setup.py]: Might be nicer to use 'with' here to handle file handle closing.
with open(filename) as f:
...     data = f.read()

**Comment by will-moore** [comment]
Install seems to be working...
```
$ pip install https://github.com/aleksandra-tarkowska/figure/tarball/pip_installer
Collecting https://github.com/aleksandra-tarkowska/figure/tarball/pip_installer
  Downloading https://github.com/aleksandra-tarkowska/figure/tarball/pip_installer (825kB)
    100% |████████████████████████████████| 829kB 687kB/s 
Requirement already satisfied (use --upgrade to upgrade): reportlab in /usr/local/lib/python2.7/site-packages (from omero-figure==1.2.2)
Requirement already satisfied (use --upgrade to upgrade): markdown in /usr/local/lib/python2.7/site-packages (from omero-figure==1.2.2)
Requirement already satisfied (use --upgrade to upgrade): pip>=1.4.1 in /usr/local/lib/python2.7/site-packages (from reportlab->omero-figure==1.2.2)
Requirement already satisfied (use --upgrade to upgrade): setuptools>=2.2 in /usr/local/lib/python2.7/site-packages (from reportlab->omero-figure==1.2.2)
Requirement already satisfied (use --upgrade to upgrade): pillow>=2.4.0 in /usr/local/lib/python2.7/site-packages (from reportlab->omero-figure==1.2.2)
Installing collected packages: omero-figure
  Running setup.py install for omero-figure ... done
Successfully installed omero-figure-1.2.2

...
>>> import figure
>>> figure.__file__
'/usr/local/lib/python2.7/site-packages/figure/__init__.pyc'
```

**Comment by joshmoore** [comment]
Should this be a `DEV` version?

**Comment by sbesson** [comment]
Quick thought: how does the script installation step work with the `pip install` workflow?

**Comment by aleksandra-tarkowska** [comment]
I think @will-moore just said that to me as well. And he feels that would be better get his PRs first merged otherwise he will have to rebase them and resolve conflict on each file. I am happy with that as long as someone merge outstanding PRs. Could we review how many of them will go it?

**Comment by aleksandra-tarkowska** [comment]
I think it will be:

```
$ mkdir test
$ cd test
$ virtuaenv venvweb
$ source venvweb/bin/activate
(venvweb) $ pip install omego omero-figure
# now instead of omero-figure is git+git://github.com/aleksandra-tarkowska/figure.git@pip_installer
(venvweb) $ omego download --ice 3.6 --branch OMERO-DEV-breaking-build py
(venvweb) $ mv OMERO.py[currentversion] OMERO.py
(venvweb) $ pip install -r OMERO.py/share/web/requirements-py27-trial.txt
(venvweb) $ OMERO.py/bin/omero config append omero.web.apps '"figure"'
(venvweb) $ OMERO.py/bin/omero config append omero.web.ui.top_links '["Figure", "figure_index", {"title": "Open Figure in new tab", "target": "figure"}]' 
(venvmapr) $ OMERO.py/bin/omero web start
(venvmapr) $ cd venvweb/lib/python2.7/site-packages/figure/scripts
(venvmapr) $ ~/test/OMERO.py/bin/omero script upload omero/figure_scripts/Figure_To_Pdf.py --official 
```

**Comment by aleksandra-tarkowska** [comment]
alternatively why just not offer that script as a zip form figure page?

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [FIGURE-merge#234](https://ci.openmicroscopy.org/job/FIGURE-merge/234/). See the [console output](https://ci.openmicroscopy.org/job/FIGURE-merge/234/consoleText) for more details.
Possible conflicts:
  - Conflict detection failed
    - static/figure/3rdparty/shape-editor-2.0.0/shape-editor.js
  - PR #165 will-moore 'Rois from omero'
    - static/figure/3rdparty/shape-editor.js

File: LICENSE [hunk 5]
```diff
@@ -0,0 +1,340 @@
+                    GNU GENERAL PUBLIC LICENSE
```
Comment (sbesson) [inline LICENSE]: This is the GPLv2 license but the package is marked as AGPLv3

Comment (aleksandra-tarkowska) [inline LICENSE]: sorry I moved the wrong file

Comment (aleksandra-tarkowska) [inline LICENSE]: https://github.com/openmicroscopy/openmicroscopy/blob/develop/components/tools/OmeroWeb/omeroweb/license/agpl-3.0.txt should be in

Comment (aleksandra-tarkowska) [inline LICENSE]: rebased now

**Comment by will-moore** [comment]
You need to update the path that ```figure/settings.py``` uses to read the ```package.json```.
```
-package_path = os.path.join(dir_path, 'package.json')
+package_path = os.path.join(dir_path, '..', 'package.json')
```

Also if you could run
```
grunt jst
```
it will update the ```templates.js``` with a change in the template ```figure/static/figure/templates/lut_picker.html``` which is currently missing from ```templates.js``` and causing bug in LUT picker.
Then commit this change to ```templates.js```.

**Comment by aleksandra-tarkowska** [comment]
I have just one extra comment about pip. I noticed that on Centos7

```
[centos@test ~]$ sudo yum install python-virtualenv
===================================================================================================================================
 Package                              Arch                      Version                           Repository                  Size
===================================================================================================================================
Installing:
 python-virtualenv                    noarch                    1.10.1-2.el7                      base                       1.2 M
Installing for dependencies:
 python-devel                         x86_64                    2.7.5-38.el7_2                    updates                    391 k
Updating for dependencies:
 python                               x86_64                    2.7.5-38.el7_2                    updates                     89 k
 python-libs                          x86_64                    2.7.5-38.el7_2                    updates                    5.6 M

[centos@test ~]$ source test/bin/activate
(test)[centos@test ~]$ pip --version
pip 1.4.1 from /home/centos/test/lib/python2.7/site-packages (python 2.7)
```

installs very old pip version. Pretty much that requires immediate `pip install --upgrade pip`

Do we want to put minim version of pip 7.0 in install_requires?

File: figure/settings.py [hunk 6]
```diff
@@ -7,8 +7,13 @@
...
 import os
-dir_path = os.path.dirname(os.path.realpath(__file__))
-package_path = os.path.join(dir_path, '..', 'package.json')
-with open(package_path) as f:
-    data = json.load(f)
-OMERO_FIGURE_VERSION = data['version']
+import pkg_resources  # part of setuptools
+
+try:
+    OMERO_FIGURE_VERSION = pkg_resources.require("omero-figure")[0].version
```
Comment (aleksandra-tarkowska) [inline figure/settings.py]: @sbesson you ok with that? Pretty much now version is only stored in `package.json`

**Comment by joshmoore** [comment]
I'd think an immediate upgrade via install_requires or even `pip install -r requirements.txt` is valid .

**Comment by will-moore** [comment]
This commit https://github.com/ome/figure/pull/168/commits/f2d70628edcfb6fcd23a0050a6b0f6e9de4d21c9 seems to not be on the various servers. E.g. http://web-dev-integration.openmicroscopy.org/static/figure/figure.js?version=1.2.2 you can still find "static/figure/templates/lut_picker.html"  instead of "figure/static/figure/templates/lut_picker.html"
so LUT dialog is broken, as reported by @pwalczysko

**Comment by will-moore** [comment]
Sorry - re: previous comment. The commit is there, but was missing the same update in the concatenated figure.js

**Comment by will-moore** [comment]
@aleksandra-tarkowska At some point, could you also open a PR to update the install instructions in index.html on the ```gh-pages-staging``` branch: https://github.com/ome/figure/blob/gh-pages-staging/index.html
Thanks.

**Comment by aleksandra-tarkowska** [comment]
@will-moore np, will do it soon

**Comment by will-moore** [comment]
That last commit has fixed the LUT dialog bug. Working fine now.

**Pull request merged**